[MDS-6011] - View geospatial bundles

matbusby-fw opened the pull request:
## Objective 

[MDS-6011](https://bcmines.atlassian.net/browse/MDS-6011)

The look of the view was already completed by @taraepp.  Previously the mapping information was stubbed in, so the work here was largely to addressing the fetching/saving of the geomark_ids and their respective mine_document_bundles.

One snag here was that we need to display the map before we have actually created the mine_docment_bundle to which we save the geomark_id, so had to do some work to accomodate different sources of that id depending on where in the process the user is.

TESTING UPDATE
I decided to add `msw` to the jest framework to simplify some kinds of tests.  I also had to upgrade a bunch of related settings packages to get it working, but it has a few nice benefits.

- Now we can mock any network response that would occur in the use of a component and specify it's return
- msw will warn you if if network request is happening that isn't mocked (and tell you exactly what it is)
- you can now interact with the dom through user click events to produce more robust tests
- if you're struggling with the data you need not being loaded immediately you can add checks for certain elements to be in the document before proceeding with further testing (ie snapshots).

I'll demonstrate the functionality when we have our testing meeting.

<img width="940" alt="image" src="https://github.com/user-attachments/assets/fe3f91b8-ebed-415a-9bd7-c05e49738636">

matbusby-fw committed:
add GEOMARK_URL_BASE to env-example

matbusby-fw committed:
Update core-api to handle the saving and fetching of mine_document_bundles in the project summary section.

matbusby-fw committed:
Update the fetching/displaying of geomark data in the project summary

taraepp commented on services/common/src/components/documents/spatial/AddSpatialDocumentsModal.tsx:
```diff
@@ -101,12 +101,29 @@ const AddSpatialDocumentsModal: FC<AddSpatialDocumentsModalProps> = ({
+      };
+    });
+    dispatch(change(modalFormName, fieldName, [...initialDocuments, ...newFiles]));
+    dispatch(change(formName, fieldName, [...initialDocuments, ...newFiles]));
```
ooooo one little maybe hiccup here-
what happens if the user uploads files, goes to review, and then cancels? That was the reason for only updating the parent form when `isFinalStep`.

matbusby-fw replied:
Hmmm tricky.  

I see what you're saying.  Since it's these documents that get sent to the viewSpatial component they needed to contain these new values.  Maybe can pass the other set of documents.

taraepp commented on services/common/src/redux/slices/spatialDataSlice.ts:
```diff
@@ -6,52 +6,90 @@ import { formatDate } from "../utils/helpers";
 import { IGeoJsonFeature } from "@mds/common/interfaces/document/geojsonFeature.interface";
 import { ISpatialBundle } from "@mds/common/interfaces/document/spatialBundle.interface";
 
+interface ItempMineDocumentWithGeomarkId extends IMineDocument {
```
I am trying to work out what an Itemp is? Typo?

matbusby-fw replied:
it's a temporary MineDocument with a geomarkId

taraepp commented on services/common/src/redux/slices/spatialDataSlice.ts:
```diff
@@ -6,52 +6,90 @@ import { formatDate } from "../utils/helpers";
-    .map((f) => {
+  const headers = createRequestHeader();
+
+  const spatial_bundles = await Promise.all(
```
This was a little helper function before, but I think now it wants to be split up into a `fetch` function and a selector to separate concerns and follow the redux pattern.

matbusby-fw replied:
This was the main issue with what took me so long.  I was trying to avoid refactoring all of your previous work since this function is providing the base for two seperate components.  One of those components just needs the file grouping and the other needs the geomark_id which sometimes needs to be fetched.

Since the state of the files (whether or not the mine_document_bundle has been created) is what dictates where it looks for the geomark_id it starts to get really janky to split up the concerns. 

I'll take another look at it.

matbusby-fw replied:
I do appreciate the eagerness, but this PR is not yet "Ready For Review"

matbusby-fw committed:
Add polyfills for Node.js globals and update document handling.

Created a new `jest.polyfills.js` for necessary Node.js globals to run Jest tests with JSDOM. Added `mine_document_bundle_id` to the `Document` model and utilized `groupSpatialBundles` in `SpatialDocumentTable`.

matbusby-fw committed:
cleaned up some functionality

matbusby-fw committed:
cleaned up some functionality

matbusby-fw committed:
updated snapshots and added msw updates to core-web

matbusby-fw committed:
update jest setups

matbusby-fw committed:
update jest configs

matbusby-fw committed:
update snap

taraepp commented on services/common/src/components/documents/spatial/AddSpatialDocumentsModal.tsx:
```diff
@@ -95,18 +99,35 @@ const AddSpatialDocumentsModal: FC<AddSpatialDocumentsModalProps> = ({
+        geomark_id,
+      };
+    });
+    dispatch(change(modalFormName, fieldName, [...initialDocuments, ...newFiles]));
```
I remember that I had to take the `initialDocuments` _out_ of the dispatch-change-modalFormName. Otherwise I'd upload a spacial bundle, save, go to upload another, and my first bundle would already be there. Have you tested it when there's already a bundle there and you're adding another?

taraepp replied:
I'm thinking that the geomark_id just wants to be a state variable or something. Then we're not having to add it to project summary documents and can keep it with the bundle.

taraepp commented on services/common/src/components/documents/spatial/AddSpatialDocumentsModal.tsx:
```diff
@@ -95,18 +99,35 @@ const AddSpatialDocumentsModal: FC<AddSpatialDocumentsModalProps> = ({
     if (isFinalStep) {
-      dispatch(change(formName, fieldName, allFiles));
+      dispatch(change(formName, fieldName, [...initialDocuments, ...newFiles]));
+      dispatch(change(formName, "spatial_documents", [...initialDocuments, ...newFiles]));
```
Was this second dispatch added by mistake?

matbusby-fw replied:
Indeed!

matbusby-fw committed:
addressed pr comments

matbusby-fw committed:
moved creation of mine_document_bundles to a helper function in the mine_document_bundle.py

simensma-fresh commented on services/common/package.json:
```diff
@@ -32,6 +32,7 @@
     "@syncfusion/ej2-react-pdfviewer": "26.1.35",
     "@syncfusion/ej2-splitbuttons": "26.1.35",
     "flagsmith": "3.19.1",
+    "msw": "2.3.5",
```
Does this need to be in `dependencies` or can it be moved to `devDependencies`?

simensma-fresh commented on services/common/package.json:
```diff
@@ -60,6 +63,7 @@
     "terser-webpack-plugin": "5.3.9",
     "ts-jest": "29.1.2",
     "typescript": "4.7.4",
+    "undici": "^6.19.7",
```
Could we lock down this one?

simensma-fresh commented on services/common/src/redux/slices/spatialDataSlice.ts:
```diff
@@ -99,19 +102,39 @@ const spatialSlice = createAppSlice({
+      async (mine_document_bundle_id: string, thunkAPI) => {
+        thunkAPI.dispatch(showLoading());
+        const headers = createRequestHeader();
+        const url = `${ENVIRONMENT.apiUrl}/mines/document-bundle/${mine_document_bundle_id}`;
```
Could we put this into `@mds/common/constants/API` together with teh rest of our endpoints?
And on that note: Any way to update the mock server to also use those constants?

matbusby-fw committed:
updated projectsummary factory to include spatial documents and document bundles

simensma-fresh commented on services/core-api/app/api/mines/documents/resources/mine_document_bundle.py:
```diff
@@ -0,0 +1,18 @@
+class MineDocumentBundleResource(Resource, UserMixin):
+    @api.doc(description='Returns list of documents associated with mines')
+    @requires_any_of([VIEW_ALL, MINESPACE_PROPONENT])
+    def get(self, mine_document_bundle_id):
```
Could we add a test for those roles to `test_expected_auth.py`?

simensma-fresh replied:
Would also be awesome to add some integration tests for this endpoint

matbusby-fw committed:
addressed more pr comments

matbusby-fw committed:
addressed more pr comments

henryoforeh-dev reviewed (dismissed):
Nice work! I intend to use the test package here in my upcoming PR too.

matbusby-fw committed:
fixed tests in api

matbusby-fw committed:
added model tests for document bundles
